internal/session/sessionwantsender.go

```diff
@@ -354,21 +362,59 @@ func (spm *sessionWantSender) processUpdates(updates []update) {
 		}
 	}
 
-	// If all available peers for a cid sent a DONT_HAVE, signal to the session
-	// that we've exhausted available peers
-	if dontHaves.Len() > 0 {
-		exhausted := spm.bpm.AllPeersDoNotHaveBlock(spm.peerAvlMgr.availablePeers(), dontHaves.Keys())
-		newlyExhausted := spm.newlyExhausted(exhausted)
-		if len(newlyExhausted) > 0 {
-			spm.onPeersExhausted(newlyExhausted)
-		}
-	}
-
 	// If any peers have sent us too many consecutive DONT_HAVEs, remove them
 	// from the session
 	for p := range prunePeers {
 		spm.SignalAvailability(p, false)
 	}
+
+	return dontHaves.Keys()
+}
+
+// checkForExhaustedWants checks if there are any wants for which all peers
+// have sent a DONT_HAVE. We call these "exhausted" wants.
+func (spm *sessionWantSender) checkForExhaustedWants(dontHaves []cid.Cid, newlyUnavailable []peer.ID) {
+	// If there are no new DONT_HAVEs, and no peers became unavailable, then
+	// we don't need to check for exhausted wants
+	if len(dontHaves) == 0 && len(newlyUnavailable) == 0 {
+		return
+	}
+
+	// We need to check each want for which we just received a DONT_HAVE
+	wants := dontHaves
+
+	// If a peer just became unavailable, then we need to check all wants
+	// (because it may be the last peer who hadn't sent a DONT_HAVE for a CID)
+	if len(newlyUnavailable) > 0 {
+		// Collect all pending wants
+		wants = make([]cid.Cid, len(spm.wants))
+		for c := range spm.wants {
+			wants = append(wants, c)
+		}
+
+		// If the last available peer in the session has become unavailable
+		// then we need to broadcast all pending wants
+		if len(spm.peerAvlMgr.availablePeers()) == 0 {
+			spm.processExhaustedWants(wants)
+			return
+		}
+	}
+
+	// If all available peers for a cid sent a DONT_HAVE, signal to the session
+	// that we've exhausted available peers
+	if len(wants) > 0 {
+		exhausted := spm.bpm.AllPeersDoNotHaveBlock(spm.peerAvlMgr.availablePeers(), wants)
+		spm.processExhaustedWants(exhausted)
+	}
+}
+
+// processExhaustedWants filters the list so that only those wants that haven't
+// already been marked as exhausted are passed to onPeersExhausted()
+func (spm *sessionWantSender) processExhaustedWants(exhausted []cid.Cid) {
+	newlyExhausted := spm.newlyExhausted(exhausted)
+	if len(newlyExhausted) > 0 {
+		go spm.onPeersExhausted(newlyExhausted)
```

Comment:
Could we comment on why this needs to be a goroutine?

Reply:
I will add a comment to the code - but just to clarify here as well:
- `Session` passes `Session.onPeersExhausted()` as a callback into `sessionWantSender`
- `Session.onPeersExhausted()` sends items on the `incoming` channel within `Session`
- The run loop in `Session` pulls items from the `incoming` channel, and the result of processing some of these items is a call to `sessionWantSender`

In summary there is a circular dependency between channels which could lead to a deadlock

Reply:
Got it. I'd be a _bit_ careful about logical races as this goroutine could be scheduled _much_ later.

I'd also consider moving the goroutine spawn into the callback itself. There, it could _try_ writing to the channel, launching a goroutine only if the write fails.

Reply:
Huh that's a clever idea, I'll do that